src/sqlfluff/dialects/dialect_cockroach.py

```diff
@@ -0,0 +1,149 @@
+"""The Cockroach dialect."""
+
+from sqlfluff.core.parser import (
+    OneOf,
+    Ref,
+    Sequence,
+    Bracketed,
+    Anything,
+    BaseSegment,
+    Delimited,
+    RegexLexer,
+    CodeSegment,
+    NamedParser,
+    RegexParser,
+    StringParser,
+    SymbolSegment,
+)
+
+from sqlfluff.core.dialects import load_raw_dialect
+
+ansi_dialect = load_raw_dialect("ansi")
+
+postgres_dialect = ansi_dialect.copy_as("postgres")
```

Comment:
This looks wrong -- it seems you're loading the `ansi` dialect, then copying it and calling it `postgres`. Did you mean to load and copy the `postgres` dialect instead?

See the Snowflake dialect -- I think it is doing what you want.

https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/dialects/dialect_snowflake.py#L31L34

Reply:
@barrywhart I have made that change, it did not change behaviour at all.

` sqlfluff lint --dialect cockroach crdb_drop_index_cascade.sql
== [crdb_drop_index_cascade.sql] FAIL
L:   1 | P:   1 |  PRS | Line 1, Position 1: Found unparsable section: 'DROP
                       | INDEX t1@i1 CASCADE'
All Finished 📜 🎉!`

The above fails, the below succeeds:

`sqlfluff lint --dialect cockroach crdb_drop_index_if_exists.sql
== [crdb_drop_index_if_exists.sql] FAIL
L:   1 | P:   6 | L010 | Inconsistent capitalisation of keywords.
All Finished 📜 🎉!`

I regenerated the .yml files, that looks the same, too.

Reply:
Can you clarify somewhere what you need help with? Please be specific, like, I want this test to work, but I'm getting <<error message>>. We can help you more effectively if that is clear. As volunteers, when maintainers have other jobs and often have just a few minutes here and there to try and help. Thanks!

Reply:
I want to lex the `@` character on its own for the Cockroach dialect. My regex sometimes works in my dialect and sometimes does not, but equally, the `::` (CAST operator) which is defined in `dialect_ansi.py` ALSO fails sometimes in the Cockroach dialect and sometimes not, I have done nothing related to casting in `dialect_cockroach.py`. The CAST operator also works correctly in the ANSI dialect. 

Is it possible that there is something amiss in `dialect_postgres.py`?

At:
```
crdb_drop_index_cascade.sql:DROP INDEX t1@i1 CASCADE;
crdb_drop_index_cascade.yml:    - raw: '@' **<---WRONG**
crdb_drop_index_if_exists.sql:DROP index IF EXISTS t1@i1;
crdb_drop_index_if_exists.yml:          at: '@' **<---CORRECT**
update_with_at_greater_than.sql:update t1 set a = jsonb_set(a, '{nata,pastryStructureEligibility,0,pastryIncluded,0}'::string[], '"PUF"') where a @> '{"nata": {"pastryStructureEligibility": [{"pastryIncluded": ["PF"]}]}}';
update_with_at_greater_than.yml:          binary_operator: '@>' **<---CORRECT (from dialect_postgres.py, nothing to do with me)**
upsert.sql:UPSERT INTO t1 (a, b, c, d, e, f, g, h, i, j, k, l, m, n, o) SELECT t2.a, t3.b, t3.c, t3.d, t3.e, t3.f, '567654', t2.h, t2.i, t2.j, t2.k, t2.l, t2.m, t3.n, t2.o FROM t2@i2 t2, t t3 WHERE t2.a = t3.a AND t2.d IS NULL AND t3.0 IS NULL AND t2.p BETWEEN 0 AND 4;
upsert.yml:    - raw: '@' **<---WRONG**
```
Casting operator:
```
❯ grep "::" * (Cockroach)
crdb_alter_table_add_column_if_not_exists_fnv.sql:ALTER TABLE t1 ADD COLUMN IF NOT EXISTS x int8 AS (fnv32(a::string, b::string, c::string)%8) stored CHECK (x IN (0, 1, 2, 3, 4, 5, 6, 7));
crdb_alter_table_add_column_if_not_exists_fnv.yml:        - raw: '::' <---WRONG
crdb_alter_table_add_column_if_not_exists_fnv.yml:        - raw: '::' <---WRONG
crdb_alter_table_add_column_if_not_exists_fnv.yml:        - raw: '::' <---WRONG
create_table_with_many_complex_constraints.sql:    CONSTRAINT c6 CHECK (n >= now()::DATE),
create_table_with_many_complex_constraints.yml:        - raw: '::' <---WRONG
update_json_remove_path.sql:UPDATE t1 SET a = json_remove_path(a, '{nata,pastryLimits,0,pastryName}'::string[]) WHERE a -> 'nata' -> 'pastryLimits' -> 0 ->> 'pastryName' NOT IN ('PUFF', 'ROLL');
update_json_remove_path.yml:                  casting_operator: '::' <---CORRECT
update_jsonb_set_jsonb_agg_jsonb_array_elements_lateral.sql:SET a = jsonb_set(a, '{inbound,rules}'::text[], t2.c)
update_jsonb_set_jsonb_agg_jsonb_array_elements_lateral.sql:                         jsonb_set(rule, '{key}'::text[], concat('"', gen_random_uuid()::text, '"')::jsonb) AS c
update_jsonb_set_jsonb_agg_jsonb_array_elements_lateral.yml:              - raw: '::' <---WRONG
update_jsonb_set_jsonb_agg_jsonb_array_elements_lateral.yml:                                      - raw: '::' <---WRONG
update_jsonb_set_jsonb_agg_jsonb_array_elements_lateral.yml:                                        - raw: '::' <---WRONG
update_jsonb_set_jsonb_agg_jsonb_array_elements_lateral.yml:                                      - raw: '::' <---WRONG
update_with_at_greater_than.sql:update t1 set a = jsonb_set(a, '{nata,pastryStructureEligibility,0,pastryIncluded,0}'::string[], '"PUF"') where a @> '{"nata": {"pastryStructureEligibility": [{"pastryIncluded": ["PF"]}]}}';
update_with_at_greater_than.yml:              - raw: '::' <---WRONG
❯ cd ../ansi
❯ grep "::" * (ANSI)
select_g.sql:    NULL::INT AS user_id,
select_g.sql:    NULL::INT AS is_paid
select_g.yml:              casting_operator: '::' <---CORRECT
select_g.yml:              casting_operator: '::' <---CORRECT
select_simple_e.sql:    my_var::date as casted_variable,
select_simple_e.sql:    123::bigint as another_casted_number
select_simple_e.yml:              casting_operator: '::' <---CORRECT
select_simple_e.yml:              casting_operator: '::' <---CORRECT
select_simple_e_nc.yml:              casting_operator: '::' <---CORRECT
select_simple_e_nc.yml:              casting_operator: '::' <---CORRECT
select_simple_h.sql:    dateadd('day', row_number() OVER (ORDER BY seq8() asc), '2014-01-01')::date AS dt
select_simple_h.yml:              casting_operator: '::' <---CORRECT
shorthand_cast.sql:    '1'    ::   INT as id1,
shorthand_cast.sql:    '2'::int as id2
shorthand_cast.yml:              casting_operator: '::' <---CORRECT
shorthand_cast.yml:              casting_operator: '::' <---CORRECT
```
I hope this makes it clearer what my problem is?

Reply:
Order is important for the lexer and parser. I wonder if you need to ensure the Cockroach additions appear before or after the related Postgres ones.

For tricky lexing or parsing issues like this, it may be helpful to first get them working in Postgres, then move them to Cockroach once you get them working. One thing at a time is often easier.

Reply:
Fair enough, I'll give that a go. Thanks!